pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,195 @@
+		memBudget *mon.BoundAccount
+		// Whenever we pass oversized message this would go to true to prevent
+		// other attempts to acquire more memory until the event goes through.
+		oversize bool
+		// If true, budget was released and no more allocations could take place.
+		closed bool
+	}
+	// Channel to notify that memory was returned to the budget.
+	replenishC chan interface{}
+
+	metrics *Metrics
+}
+
+func (f *feedBudget) init(budget *mon.BoundAccount, metrics *Metrics) {
+	f.mu.memBudget = budget
+	if budget != nil {
+		f.replenishC = make(chan interface{}, 1)
+		f.metrics = metrics
+	}
+}
+
+// Get allocates amount from budget.
+// If budget has less memory than requested it could either wait, if we have any
+// in flight allocations, or it could send a single message through regardless
+// of its size. In latter case, memory is not requested from underlying budget,
+// but we won't be able to do another Get until this message is returned. To
+// track such events, feedBudget uses a counter metric.
+//
+// timeout argument indicates how long we could wait for the budget to replenish
+// if there is not enough budget available. Budget is replenished by Return
+// calls.
+// timeout value of 0 means wait indefinitely.
+// Whenever we use timeout, we return its wait chan to caller, so it could be
+// used subsequently by the chained wait for other resources thus keeping total
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *feedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration, stoppedC <-chan struct{},
+) (sharedBudgetAllocation, <-chan time.Time, error) {
+	// Function to allocate from budget or if there's nothing there borrow from
+	// emergency ether.
+	tryAlloc := func() (sharedBudgetAllocation, error) {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+		if f.mu.closed {
+			return sharedBudgetAllocation{}, errors.New("budget is already released")
+		}
+		if f.mu.oversize {
+			return sharedBudgetAllocation{}, errors.New("failed to allocate mem budget")
+		}
+		err := f.mu.memBudget.Grow(ctx, amount)
+		if err == nil {
+			return sharedBudgetAllocation{size: amount, refCount: 1}, nil
+		}
+		// If we didn't use any budget yet and not currently sending an oversized
+		// event then we can send one.
+		if f.mu.memBudget.Used() == 0 {
+			f.metrics.RangeFeedOverBudgetEvents.Inc(1)
+			f.mu.oversize = true
+			// We are in override mode, we return 0 size to avoid returning it to
+			// budget later as we didn't borrow anything.
+			return sharedBudgetAllocation{size: 0, refCount: 1}, nil
+		}
+		return sharedBudgetAllocation{}, err
+	}
+
+	alloc, err := tryAlloc()
+	if err == nil {
+		return alloc, nil, nil
+	}
+	// If we failed to allocate immediately, then we need to wait on replenish
+	// channel to track Return calls from consumed events.
+	if timeout == 0 {
```

Comment:
I think you can combine this if with the else portion.
```
var timeC <-chan time.Time
if timeout > 0 {
   timeC = ....
}
```

Reading from nil channel blocks forever.